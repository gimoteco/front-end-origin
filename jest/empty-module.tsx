import * as React from 'react';
const SvgrMock = () => <span />;

export default SvgrMock;

export const ReactComponent = SvgrMock;
