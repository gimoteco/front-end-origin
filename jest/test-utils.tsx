import { render } from '@testing-library/react';
import * as React from 'react';
import { ThemeProvider } from 'styled-components';
import theme from '../src/styles/theme';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';
// override render method
export { customRender as render };
