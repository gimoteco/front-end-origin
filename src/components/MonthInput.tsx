import { addMonths, format as formatDate } from 'date-fns';
import * as React from 'react';
import Arrow from '../icons/arrow.svg';
import * as Style from './MonthInput.styles';
function MonthInput() {
  // const [month, year] = ['October', 2020];
  const [monthDate, setMonth] = React.useState(new Date());
  const month = formatDate(monthDate, 'LLLL');
  const year = formatDate(monthDate, 'yyyy');

  const goToPreviousMonth = React.useCallback(() => {
    setMonth(addMonths(monthDate, -1));
  }, [monthDate, setMonth]);
  const goToNextMonth = React.useCallback(() => {
    setMonth(addMonths(monthDate, 1));
  }, [monthDate, setMonth]);

  return (
    <Style.MonthInputWrapper>
      <Style.StepButton onClick={goToPreviousMonth}>
        <Arrow />
      </Style.StepButton>
      <Style.MonthIndicator>
        <Style.Month>{month}</Style.Month>
        <Style.Year>{year}</Style.Year>
      </Style.MonthIndicator>

      <Style.StepButton onClick={goToNextMonth}>
        <Arrow />
      </Style.StepButton>
    </Style.MonthInputWrapper>
  );
}

export default React.memo(MonthInput);
