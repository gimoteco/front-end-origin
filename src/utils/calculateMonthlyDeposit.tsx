import {
  addDays,
  differenceInMonths,
  endOfMonth,
  startOfMonth
} from 'date-fns';

export function calculateMonthlyDeposit(total: number, reachDate: Date) {
  const thisMonth = startOfMonth(new Date());
  const endMonth = addDays(endOfMonth(reachDate), 1);
  const numberOfDeposits = differenceInMonths(endMonth, thisMonth);
  const depositValue = total / numberOfDeposits;
  return { numberOfDeposits, depositValue };
}
