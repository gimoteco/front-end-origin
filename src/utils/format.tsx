import { differenceInMonths } from 'date-fns';

const formatter = new Intl.NumberFormat('en-US', {
  style: 'currency',
  currency: 'USD',
  minimumFractionDigits: 0
});

export function formatMoney(value: number) {
  return formatter.format(value);
}

export function calculateMonthlyDeposit(total: number, reachDate: Date) {
  const today = new Date();
  const numberOfDeposits = differenceInMonths(reachDate, today);
  const depositValue = total / numberOfDeposits;
  return { numberOfDeposits, depositValue };
}
