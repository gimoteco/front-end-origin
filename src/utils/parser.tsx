import { format as formatDate } from 'date-fns';

export function parseMoney(currency: string) {
  return Number(currency.replace(/[^0-9.-]+/g, ''));
}

export function parseDateParts(date: Date) {
  const month = formatDate(date, 'LLLL');
  const year = formatDate(date, 'yyyy');
  return { month, year };
}
