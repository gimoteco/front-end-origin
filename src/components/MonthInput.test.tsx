import { addMonths } from 'date-fns';
import * as React from 'react';
import { fireEvent, render } from '../../jest/test-utils';
import { parseDateParts } from '../utils/parser';
import MonthInput from './MonthInput';

describe('MonthInput', () => {
  it('should display the month value', async () => {
    const date = new Date(2020, 0, 1);
    const { month, year } = parseDateParts(date);

    const { findByText } = render(
      <MonthInput id="some-id" value={date} onChange={jest.fn()} />
    );

    expect(await findByText(month)).toBeInTheDocument();
    expect(await findByText(year)).toBeInTheDocument();
  });

  it('should go to the next month', async () => {
    const date = new Date(2020, 0, 1);
    const { month: nextMonth, year: nextMonthYear } = parseDateParts(
      addMonths(date, 1)
    );

    const { findAllByRole, findByText } = render(
      <MonthInput id="some-id" value={date} onChange={jest.fn()} />
    );

    fireEvent.click((await findAllByRole('button'))[1]);

    expect(await findByText(nextMonth)).toBeInTheDocument();
    expect(await findByText(nextMonthYear)).toBeInTheDocument();
  });

  it('should go to the previous month', async () => {
    const date = new Date(2050, 0, 1);
    const { month: previousMonth, year: previousMonthYear } = parseDateParts(
      addMonths(date, -1)
    );

    const { findAllByRole, findByText } = render(
      <MonthInput id="some-id" value={date} onChange={jest.fn()} />
    );

    fireEvent.click((await findAllByRole('button'))[0]);

    expect(await findByText(previousMonth)).toBeInTheDocument();
    expect(await findByText(previousMonthYear)).toBeInTheDocument();
  });

  it('shouldn`t go to the previous month if is past', async () => {
    const date = new Date();
    const { month: currentMonth, year: currentYear } = parseDateParts(date);
    const { month: previousMonth } = parseDateParts(addMonths(date, -1));

    const { findAllByRole, findByText, queryAllByText } = render(
      <MonthInput id="some-id" value={date} onChange={jest.fn()} />
    );

    fireEvent.click((await findAllByRole('button'))[0]);

    expect(await queryAllByText(previousMonth)).toEqual([]);
    expect(await findByText(currentYear)).toBeInTheDocument();
    expect(await findByText(currentMonth)).toBeInTheDocument();
  });

  it('should navigate to the next month through keyboard right arrow', async () => {
    const date = new Date();
    const id = 'some-id';
    const { month: nextMonth, year: nextMonthYear } = parseDateParts(
      addMonths(date, 1)
    );
    const { getByTestId, findByText } = render(
      <MonthInput id={id} value={date} onChange={jest.fn()} />
    );
    const input = await getByTestId(id);
    console.log(date);

    fireEvent.keyDown(input, {
      key: 'ArrowRight',
      code: 'ArrowRight',
      which: 39
    });

    expect(await findByText(nextMonthYear)).toBeInTheDocument();
    expect(await findByText(nextMonth)).toBeInTheDocument();
  });

  it('should navigate to the previous month through keyboard left arrow', async () => {
    const date = addMonths(new Date(), 5);
    const { month: previousMonth, year: previousMonthYear } = parseDateParts(
      addMonths(date, -1)
    );
    const id = 'some-id';
    const { getByTestId, findByText } = render(
      <MonthInput id={id} value={date} onChange={jest.fn()} />
    );
    const input = getByTestId(id);

    fireEvent.keyDown(input, {
      key: 'ArrowLeft',
      code: 'ArrowLeft',
      which: 37
    });

    expect(await findByText(previousMonth)).toBeInTheDocument();
    expect(await findByText(previousMonthYear)).toBeInTheDocument();
  });
});
