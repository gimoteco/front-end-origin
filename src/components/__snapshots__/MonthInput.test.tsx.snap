// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MonthInput should render according snapshot 1`] = `
<div>
  <div
    class="sc-bwzfXH iXwpvX"
    data-testid="some-id"
    id="some-id"
    tabindex="0"
  >
    <button
      class="sc-bdVaJa bWNBEi"
      disabled=""
      type="button"
    >
      <span />
    </button>
    <div
      class="sc-htpNat gSEWri"
    >
      <span
        class="sc-bxivhb AHobs"
      >
        January
      </span>
      <span
        class="sc-ifAKCX kqPTaU"
      >
        2000
      </span>
    </div>
    <button
      class="sc-bdVaJa bWNBEi"
      type="button"
    >
      <span />
    </button>
  </div>
</div>
`;
